cf-monitord/mon_network.c

```diff
@@ -202,42 +218,292 @@ void MonNetworkGatherData(double *cf_this)
+
+    if (state == SOCK_STATE_LISTEN)
+    {
+        char port_str[6];       /* max "65536" (plus '\0') */
+        snprintf(port_str, sizeof(port_str), "%d", local_port);
+
+        IdempPrependItem(&ALL_INCOMING, port_str, NULL);
+
+        switch (type)
+        {
+        case cfn_tcp4:
+            IdempPrependItem(&MON_TCP4, port_str, local_addr);
+            break;
+        case cfn_tcp6:
+            IdempPrependItem(&MON_TCP6, port_str, local_addr);
+            break;
+        case cfn_udp4:
+            IdempPrependItem(&MON_UDP4, port_str, local_addr);
+            break;
+        case cfn_udp6:
+            IdempPrependItem(&MON_UDP6, port_str, local_addr);
+            break;
+        default:
+            debug_abort_if_reached();
+            break;
+        }
+    }
+    for (size_t i = 0; i < ATTR; i++)
+    {
+        if (local_port == ECGSOCKS[i].port)
+        {
+            cf_this[ECGSOCKS[i].in]++;
+            AppendItem(&in[i], socket_info, "");
+
+        }
+
+        if (remote_port == ECGSOCKS[i].port)
+        {
+            cf_this[ECGSOCKS[i].out]++;
+            AppendItem(&out[i], socket_info, "");
+
+        }
+    }
+}
+
+static bool GetNetworkDataFromProcNet(double *cf_this, Item **in, Item **out)
+{
+    FILE *fp = fopen("/proc/net/tcp", "r");
+    if (fp == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to open /proc/net/tcp for reading");
+        return false;
+    }
+
+    bool result = true;
+    Seq *lines = NULL;
+
+    size_t buff_size = 256;
+    char *buff = xmalloc(buff_size);
+
+    /* Read the header */
+    ssize_t ret = CfReadLine(&buff, &buff_size, fp);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read data from /proc/net/tcp");
+        result = false;
+        goto cleanup;
+    }
+
+    /* Read real data */
+    lines = SeqNew(64, free);
+    ret = CfReadLines(&buff, &buff_size, fp, lines);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read data from /proc/net/tcp");
+        result = false;
+        goto cleanup;
+    }
+    Log(LOG_LEVEL_VERBOSE, "Read %zu lines from /proc/net/tcp", ret);
+
+    /* else */
+    char local_addr[INET_ADDRSTRLEN];
+    char remote_addr[INET_ADDRSTRLEN];
+    uint32_t l_port, r_port;
+    SocketState state;
+    for (size_t i = 0; i < ret; i++)
+    {
+        char *line = SeqAt(lines,i);
+        if (ParseIPv4SocketInfo(line, local_addr, &l_port, remote_addr, &r_port, &state))
+        {
+            SaveSocketInfo(local_addr, l_port, r_port,
+                           state, cfn_tcp4,
+                           line, cf_this, in, out);
+        }
+    }
+    SeqClear(lines);
+
+    char local_addr6[INET6_ADDRSTRLEN];
+    char remote_addr6[INET6_ADDRSTRLEN];
+    fp = fopen("/proc/net/tcp6", "r");
+    if (fp == NULL)
+    {
+        /* IPv6 may be completely disabled in kernel so this should be handled gracefully. */
+        Log(LOG_LEVEL_VERBOSE, "Failed to read data from /proc/net/tcp6");
+    }
+    else
```

Comment:
If you made separate (inline) functions for the different files, error handling would be more clear, less `if`/ `else`.